core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -722,6 +735,251 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
+
+    /**
+     * Executes a sequence by invoking in a blocking fashion each of its components.
+     *
+     * @param user the user invoking the action
+     * @param action the sequence action to be invoked
+     * @param env the merged parameters from the package/reference if any TODO: check ignoring these params
+     * @param payload the dynamic arguments for the activation
+     * @param blocking true iff this is a blocking invoke
+     * @param components the actions in the sequence
+     * @param cause the id of the activation that caused this sequence (defined only for inner sequences and None for topmost sequences)
+     * @param transid a transaction id for logging
+     * @return a future of type (ActivationId, Some(WhiskActivation)) if blocking; else (ActivationId, None)
+     */
+    private def invokeSequence(
+            user: Identity,
+            action: WhiskAction,
+            //env: Option[Parameters],
+            payload: Option[JsObject],
+            blocking: Boolean,
+            topmost: Boolean,
+            components: Vector[FullyQualifiedEntityName],
+            cause: Option[ActivationId])(implicit transid: TransactionId): Future[(ActivationId, Option[WhiskActivation])] = {
+        // create new activation id that corresponds to the sequence
+        val seqActivationId = activationIdFactory.make()
+        info(this, s"Invoking sequence $action topmost $topmost activationid '$seqActivationId'")
+        // the cause for the component activations is the current sequence
+        val futureWskActivations = invokeSequenceComponents(user, action, seqActivationId, payload, components, cause = Some(seqActivationId))
+        val futureSeqResult = Future.sequence(futureWskActivations)
+        val response: Future[(ActivationId, Option[WhiskActivation])] =
+            if (topmost) { // need to deal with blocking and closing connection
+                if (blocking) {
+                    val timeout = maxWaitForBlockingActivation + blockingInvokeGrace
+                    val futureSeqResultTimeout = futureSeqResult withTimeout(timeout, new BlockingInvokeTimeout(seqActivationId))
+                    futureSeqResultTimeout map { wskActivations =>
+                        // the execution of the sequence was successful, return the result
+                        val seqActivation = makeSeqActivation(user, action, seqActivationId, wskActivations, topmost, wskActivations.last.response, cause)
+                        (seqActivationId, Some(seqActivation))
+                    }
+                } else {
+                    // non-blocking sequence execution, return activation id
+                    Future.successful((seqActivationId, None))
+                }
+            } else {
+                // not topmost, blocking for sure, wait for the "aggregate" future to finish (no timeout)
+                val seqResult = Await.ready(futureSeqResult, Duration.Inf).value.get
```

Comment:
is there a way to avoid using Await? this blocks the thread and is not guaranteed to be deadlock free. also the Duration.Inf is worrisome.